Python/optimizer_analysis.c

```diff
@@ -9,10 +20,532 @@
+    if (frame == NULL) {
+        return -1;
+    }
+    ctx->curr_frame_depth++;
+    ctx->frame = frame;
+
+    // IR and sym setup
+    ctx->frequent_syms.push_nulL_sym = NULL;
+
+    // Emitter setup
+    ctx->emitter.writebuffer = new_writebuffer;
+    ctx->emitter.curr_i = 0;
+    ctx->emitter.writebuffer_end = new_writebuffer + ir_entries;
+    return 0;
+}
+
+
+static int
+ctx_frame_pop(
+    _Py_UOpsAbstractInterpContext *ctx
+)
+{
+    _Py_UOpsAbstractFrame *frame = ctx->frame;
+
+    ctx->n_consumed = frame->locals;
+    ctx->curr_frame_depth--;
+    assert(ctx->curr_frame_depth >= 1);
+    ctx->frame = &ctx->frames[ctx->curr_frame_depth - 1];
+
+    return 0;
+}
+
+static void
+sym_set_type_from_const(_Py_UOpsSymType *sym, PyObject *obj);
+
+// Steals a reference to const_val
+static _Py_UOpsSymType*
+_Py_UOpsSymType_New(_Py_UOpsAbstractInterpContext *ctx,
+                               PyObject *const_val)
+{
+    _Py_UOpsSymType *self = &ctx->t_arena.arena[ctx->t_arena.ty_curr_number];
+    if (ctx->t_arena.ty_curr_number >= ctx->t_arena.ty_max_number) {
+        OPT_STAT_INC(optimizer_failure_reason_no_memory);
+        DPRINTF(1, "out of space for symbolic expression type\n");
+        return NULL;
+    }
+    ctx->t_arena.ty_curr_number++;
+    self->const_val = NULL;
+    self->types = 0;
+
+    if (const_val != NULL) {
+        self->const_val = Py_NewRef(const_val);
+    }
+
+    return self;
+}
+
+
+static void
+sym_set_type(_Py_UOpsSymType *sym, _Py_UOpsSymExprTypeEnum typ)
+{
+    sym->types |= 1 << typ;
+}
+
+
+static _Py_UOpsSymExprTypeEnum
+pytype_to_type(PyTypeObject *tp)
+{
+    _Py_UOpsSymExprTypeEnum typ = INVALID_TYPE;
+    if (tp == NULL) {
+        typ = NULL_TYPE;
+    }
+    else if (tp == &PyLong_Type) {
+        typ = PYLONG_TYPE;
+    }
+    else if (tp == &PyFloat_Type) {
+        typ = PYFLOAT_TYPE;
+    }
+    else if (tp == &PyUnicode_Type) {
+        typ = PYUNICODE_TYPE;
+    }
+    else if (tp == &PyFunction_Type) {
+        typ = PYFUNCTION_TYPE_VERSION_TYPE;
+    }
+    else if (tp == &PyMethod_Type) {
+        typ = PYMETHOD_TYPE;
+    }
+    return typ;
+}
+
+static void
+sym_set_pytype(_Py_UOpsSymType *sym, PyTypeObject *tp)
+{
+    assert(tp == NULL || PyType_Check(tp));
+    sym_set_type(sym, pytype_to_type(tp));
+}
+
+static void
+sym_set_type_from_const(_Py_UOpsSymType *sym, PyObject *obj)
+{
+    PyTypeObject *tp = Py_TYPE(obj);
+
+    if (tp == &PyFunction_Type) {
+        sym_set_type(sym, PYFUNCTION_TYPE_VERSION_TYPE);
+    }
+    else {
+        sym_set_pytype(sym, tp);
+    }
+
+}
+
+static inline _Py_UOpsSymType*
+sym_init_unknown(_Py_UOpsAbstractInterpContext *ctx)
+{
+    return _Py_UOpsSymType_New(ctx,NULL);
+}
+
+static inline bool
+sym_is_unknown_type(_Py_UOpsSymType *typ)
+{
+    return (typ->types == 0) || (typ->types == (1U << INVALID_TYPE));
+}
+
+static inline _Py_UOpsSymType*
+sym_init_known_type(_Py_UOpsAbstractInterpContext *ctx,
```

Comment:
What is this? It is not bottom, that is `sym_init_unknown()`, I think.
So is it not-NULL? As it has a type (so can't be NULL), but we don't know what it is?